Libraries/LibTLS/Exchange.cpp

```diff
@@ -0,0 +1,216 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibCore/Timer.h>
+#include <LibCrypto/ASN1/DER.h>
+#include <LibCrypto/PK/Code/EMSA_PSS.h>
+#include <LibTLS/TLSv12.h>
+
+namespace TLS {
+
+bool TLSv12::expand_key()
+{
+    u8 key[192]; // soooooooo many constants
+    auto key_buffer = ByteBuffer::wrap(key, 192);
+
+    if (m_context.master_key.size() == 0) {
+        dbg() << "expand_key() with empty master key";
+        return false;
+    }
+
+    auto key_size = key_length();
+    auto mac_size = mac_length();
+    auto iv_size = iv_length();
+
+    pseudorandom_function(
+        key_buffer,
+        m_context.master_key,
+        (const u8*)"key expansion", 13,
+        ByteBuffer::wrap(m_context.remote_random, 32),
+        ByteBuffer::wrap(m_context.local_random, 32));
+
+    size_t offset = 0;
+    memcpy(m_context.crypto.local_mac, key + offset, mac_size);
+    offset += mac_size;
+    memcpy(m_context.crypto.remote_mac, key + offset, mac_size);
+    offset += mac_size;
+
+    auto client_key = key + offset;
+    offset += key_size;
+    auto server_key = key + offset;
+    offset += key_size;
+    auto client_iv = key + offset;
+    offset += iv_size;
+    auto server_iv = key + offset;
+    offset += iv_size;
+
+#ifdef TLS_DEBUG
+    dbg() << "client key";
+    print_buffer(client_key, key_size);
+    dbg() << "server key";
+    print_buffer(server_key, key_size);
+    dbg() << "client iv";
+    print_buffer(client_iv, iv_size);
+    dbg() << "server iv";
+    print_buffer(server_iv, iv_size);
+    dbg() << "client mac key";
+    print_buffer(m_context.crypto.local_mac, mac_size);
+    dbg() << "server mac key";
+    print_buffer(m_context.crypto.remote_mac, mac_size);
+#endif
+
+    memcpy(m_context.crypto.local_iv, client_iv, iv_size);
+    memcpy(m_context.crypto.remote_iv, server_iv, iv_size);
+
+    m_aes_local = make<Crypto::Cipher::AESCipher::CBCMode>(ByteBuffer::wrap(client_key, key_size), key_size * 8, Crypto::Cipher::Intent::Encryption);
+    m_aes_remote = make<Crypto::Cipher::AESCipher::CBCMode>(ByteBuffer::wrap(server_key, key_size), key_size * 8, Crypto::Cipher::Intent::Decryption);
+
+    m_context.crypto.created = 1;
+
+    return true;
+}
+
+void TLSv12::pseudorandom_function(ByteBuffer& output, const ByteBuffer& secret, const u8* label, size_t label_length, const ByteBuffer& seed, const ByteBuffer& seed_b)
+{
+    if (!secret.size()) {
+        dbg() << "null secret";
+        return;
+    }
+
+    // RFC 5246: "In this section, we define one PRF, based on HMAC.  This PRF with the
+    // 			  SHA-256 hash function is used for all cipher suites defined in this
+    //            document and in TLS documents published prior to this document when
+    //            TLS 1.2 is negotiated."
```

Comment:
Styling : Formatting issue in this comment ?

Reply:
oh...I blame RFC texts using hard tabs